application/src/main/java/org/opentripplanner/updater/trip/metrics/StreamingTripUpdateMetrics.java

```diff
@@ -25,66 +23,58 @@
 public class StreamingTripUpdateMetrics extends TripUpdateMetrics {
 
   protected static final String METRICS_PREFIX = "streaming_trip_updates";
-  private final Counter successfulCounter;
-  private final Counter failureCounter;
-  private final Counter warningsCounter;
-  private final Map<UpdateError.UpdateErrorType, Counter> failuresByType = new HashMap<>();
-  private final Map<UpdateSuccess.WarningType, Counter> warningsByType = new HashMap<>();
+  private final boolean producerMetrics;
 
   public StreamingTripUpdateMetrics(UrlUpdaterParameters parameters) {
     super(parameters);
-    this.successfulCounter = getCounter("successful", "Total successfully applied trip updates");
-    this.failureCounter = getCounter("failed", "Total failed trip updates");
-    this.warningsCounter = getCounter("warnings", "Total warnings for successful trip updates");
+    this.producerMetrics = parameters.producerMetrics();
   }
 
   public void setCounters(UpdateResult result) {
-    this.successfulCounter.increment(result.successful());
-    this.failureCounter.increment(result.failed());
-    this.warningsCounter.increment(result.warnings().size());
-
-    setFailures(result);
-    setWarnings(result);
+    incrementWarningCounts(result);
+    incrementFailureCounts(result);
+    incrementSuccessCounts(result);
   }
 
-  private void setWarnings(UpdateResult result) {
+  private void incrementWarningCounts(UpdateResult result) {
     for (var warningType : result.warnings()) {
-      var counter = warningsByType.get(warningType);
-      if (Objects.isNull(counter)) {
-        counter =
-          getCounter(
-            "warning_type",
-            "Total warnings by type generated by successful trip updates",
-            Tag.of("warningType", warningType.name())
-          );
-        warningsByType.put(warningType, counter);
-      }
-      counter.increment();
+      Tags tags = Tags.concat(baseTags, Tags.of("warningType", warningType.name()));
+      Counter
+        .builder(METRICS_PREFIX + "." + "warnings")
+        .description("Total warnings by type generated by successful trip updates")
+        .tags(tags)
+        .register(Metrics.globalRegistry)
+        .increment();
     }
   }
 
-  private void setFailures(UpdateResult result) {
-    for (var errorType : result.failures().keySet()) {
-      var counter = failuresByType.get(errorType);
-      if (Objects.isNull(counter)) {
-        counter =
-          getCounter(
-            "failure_type",
-            "Total failed trip updates by type",
-            Tag.of("errorType", errorType.name())
-          );
-        failuresByType.put(errorType, counter);
+  private void incrementFailureCounts(UpdateResult result) {
+    for (UpdateError error : result.errors()) {
+      Tags tags = Tags.concat(baseTags, Tags.of("errorType", error.errorType().name()));
+      if (producerMetrics) {
+        tags = tags.and(Tag.of("producer", error.producer()));
```

Comment:
Are you allow to put null as the tag value? I think you should check if there really is a producer set.

Reply:
Good catch. Thanks!